File: WMS-Client/src/app/services/attendance.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../environments/environment';
import { Observable } from 'rxjs';

export interface AttendanceDto {
  id: number;
  employeeId: number;
  employeeName?: string;
  attendanceDate: string;
  checkIn: string | null;
  checkOut: string | null;
  status: string;
  remarks?: string | null;
}

export interface TodayActiveEmployeeDto {
  attendanceId: number;
  employeeId: number;
  employeeName: string;
  employeeCode: string;
  departmentName: string;
  checkInTime: string;
}

export interface TodayActiveDto {
  totalActive: number;
  employees: TodayActiveEmployeeDto[];
}

@Injectable({ providedIn: 'root' })
export class AttendanceService {
  constructor(private readonly http: HttpClient) {}

  getMonthly(employeeId: number, year: number, month: number): Observable<AttendanceDto[]> {
    return this.http.get<AttendanceDto[]>(`${environment.apiUrl}/attendance/employee/${employeeId}/month/${year}/${month}`);
  }

  checkIn(request: { employeeId: number; attendanceDate?: string; checkInTime?: string }) {
    return this.http.post(`${environment.apiUrl}/attendance/check-in`, request);
  }

  checkOut(id: number, request: { checkOutTime?: string; remarks?: string }) {
    return this.http.post(`${environment.apiUrl}/attendance/${id}/check-out`, request);
  }

  getAll(): Observable<AttendanceDto[]> {
    return this.http.get<AttendanceDto[]>(`${environment.apiUrl}/attendance/all`);
  }

  getTodayActive(): Observable<TodayActiveDto> {
    return this.http.get<TodayActiveDto>(`${environment.apiUrl}/attendance/today-active`);
  }
}

File: WMS.Application/Services/AttendanceService.cs
using AutoMapper;
using WMS.Application.Interfaces.Repositories;
using WMS.Application.Interfaces.Services;
using WMS.Application.Models.Attendance;
using WMS.Domain.Entities;
using WMS.Domain.Enums;

namespace WMS.Application.Services;

public class AttendanceService(IUnitOfWork unitOfWork, IMapper mapper) : IAttendanceService
{
    public async Task<IReadOnlyList<AttendanceDto>> GetByEmployeeAndMonthAsync(int employeeId, int year, int month, CancellationToken cancellationToken = default)
    {
        var attendances = await unitOfWork.Attendances.GetByEmployeeAndMonthAsync(employeeId, year, month, cancellationToken);
        return mapper.Map<IReadOnlyList<AttendanceDto>>(attendances);
    }

    public async Task<int> CheckInAsync(CheckInDto request, CancellationToken cancellationToken = default)
    {
        var attendance = new Attendance
        {
            EmployeeId = request.EmployeeId,
            AttendanceDate = request.AttendanceDate ?? DateTime.UtcNow.Date,
            CheckInTime = request.CheckInTime ?? DateTime.UtcNow.TimeOfDay,
            Status = AttendanceStatus.Present
        };

        await unitOfWork.Attendances.AddAsync(attendance, cancellationToken);
        await unitOfWork.SaveChangesAsync(cancellationToken);
        return attendance.Id;
    }

    public async Task CheckOutAsync(int attendanceId, CheckOutDto request, CancellationToken cancellationToken = default)
    {
        var attendance = await unitOfWork.Attendances.GetByIdAsync(attendanceId, cancellationToken)
            ?? throw new KeyNotFoundException($"Attendance {attendanceId} was not found.");

        attendance.CheckOutTime = request.CheckOutTime ?? DateTime.UtcNow.TimeOfDay;
        attendance.Remarks = request.Remarks;
        unitOfWork.Attendances.Update(attendance);
        await unitOfWork.SaveChangesAsync(cancellationToken);
    }

    public async Task<TodayActiveDto> GetTodayActiveAsync(CancellationToken cancellationToken = default)
    {
        var activeAttendances = await unitOfWork.Attendances.GetTodayActiveAsync(cancellationToken);
        return new TodayActiveDto
        {
            TotalActive = activeAttendances.Count,
            Employees = activeAttendances.Select(a => new TodayActiveEmployeeDto
            {
                AttendanceId = a.Id,
                EmployeeId = a.EmployeeId,
                EmployeeName = a.Employee!.FullName,
                EmployeeCode = a.Employee.EmployeeCode,
                DepartmentName = a.Employee.Department?.Name ?? "",
                CheckInTime = a.CheckInTime.ToString(@"hh\:mm")
            }).ToList()
        };
    }
}